File: ch5/5-5.c
#include <stdio.h>

char *strn_cpy(char *d, char *s, int n);
char *strn_cat(char *d, char *s, int n);
int strn_cmp(char *s, char *t, int n);

int main(void)
{
    char *s1 = "john";
    char *s2 = "johnny";
    char s3[20];

    printf("%d\n", strn_cmp(s1, s2, 2));
    printf("%s\n", strn_cpy(s3, s1, 20));
    printf("%s\n", strn_cat(s3, s2, 20));
    return 0;
}

char *strn_cpy(char *d, char *s, int n)
{
    int i;
    char *r;
    r = d;
    for (i = 0; i < n && *s != '\0'; i++) {
        *d++ = *s++;
    }
    *d = '\0';
    return r;
}

char *strn_cat(char *d, char *s, int n)
{
    char *r;
    int i;
    r = d;
    for (; *d != '\0'; d++) {
        ;
    }
    for (i = 0; *s != '\0' && i < n; i++) {
        *d++ = *s++;
    }
    *d = '\0';
    return r;
}

int strn_cmp(char *s, char *t, int n)
{
    int i;
    for (i = 0; i < n && (*s++ == *t++); i++) {
        ;
    }
    if (i == n) {
        return 0;
    } else {
        return (*(s-1) > *(t-1)) ? 1 : -1;
    }
}

File: ch5/5-5.go
package main

import "fmt"

func main() {
	s1 := []byte("john")
	s2 := []byte("johnny")
	s3 := make([]byte, 20)

	fmt.Printf("%d\n", strn_cmp(s1, s2, 2))
	fmt.Printf("%s\n", string(strn_cpy(s3, s1, 20)))
	fmt.Printf("%s\n", string(strn_cat(s3, s2, 20)))
}

func strn_cpy(d []byte, s []byte, n int) []byte {
	var i int
	var r []byte
	r = d
	for i = 0; i < n && i < len(s); i++ {
		d[i] = s[i]
	}
	return r
}

func strn_cat(d []byte, s []byte, n int) []byte {
	var r []byte
	var i, j int
	r = d
	for j = 0; d[j] != 0; j++ {
		// nothing
	}
	for i = 0; i < len(s) && i < n; i, j = i+1, j+1 {
		d[j] = s[i]
	}
	return r
}

func strn_cmp(s []byte, t []byte, n int) int {
	var i int
	for i = 0; i < n && (s[i] == t[i]); i++ {
		// nothing
	}
	if i == n {
		return 0
	} else {
		if s[i-1] > t[i-1] {
			return 1
		} else {
			return -1
		}
	}
}
